pkg/slo-controller/noderesource/plugins/batchresource/util.go

```diff
@@ -0,0 +1,318 @@
+		batchAllocatable.Cpu().MilliValue(), nodeAllocatable.Cpu().MilliValue(), nodeReserve.Cpu().MilliValue(),
+		systemUsed.Cpu().MilliValue(), podHPUsed.Cpu().MilliValue())
+
+	var memMsg string
+	if strategy != nil && strategy.MemoryCalculatePolicy != nil && *strategy.MemoryCalculatePolicy == configuration.CalculateByPodRequest {
+		batchAllocatable[corev1.ResourceMemory] = *batchAllocatableByRequest.Memory()
+		memMsg = fmt.Sprintf("batchAllocatable[Mem(GB)]:%v = nodeAllocatable:%v - nodeReservation:%v - podHPRequest:%v",
+			batchAllocatable.Memory().ScaledValue(resource.Giga), nodeAllocatable.Memory().ScaledValue(resource.Giga),
+			nodeReserve.Memory().ScaledValue(resource.Giga), podHPReq.Memory().ScaledValue(resource.Giga))
+	} else { // use CalculatePolicy "usage" by default
+		memMsg = fmt.Sprintf("batchAllocatable[Mem(GB)]:%v = nodeAllocatable:%v - nodeReservation:%v - systemUsage:%v - podHPUsed:%v",
+			batchAllocatable.Memory().ScaledValue(resource.Giga), nodeAllocatable.Memory().ScaledValue(resource.Giga),
+			nodeReserve.Memory().ScaledValue(resource.Giga), systemUsed.Memory().ScaledValue(resource.Giga),
+			podHPUsed.Memory().ScaledValue(resource.Giga))
+	}
+
+	return batchAllocatable, cpuMsg, memMsg
+}
+
+func prepareNodeForResource(node *corev1.Node, nr *framework.NodeResource, name corev1.ResourceName) {
+	q := nr.Resources[name]
+	if q == nil || nr.Resets[name] { // if the specified resource has no quantity
+		delete(node.Status.Capacity, name)
+		delete(node.Status.Allocatable, name)
+		return
+	}
+
+	// amplify batch cpu according to cpu normalization ratio
+	if name == extension.BatchCPU {
+		ratio, err := getCPUNormalizationRatio(nr)
+		if err != nil {
+			klog.V(5).InfoS("failed to get cpu normalization ratio for node extended resources",
+				"node", node.Name, "err", err)
+		}
+		if ratio > 1.0 { // skip for invalid ratio
+			newQuantity := util.MultiplyMilliQuant(*q, ratio)
+			q = &newQuantity
+		}
+	}
+
+	// NOTE: extended resource would be validated as an integer, so it should be checked before the update
+	if _, ok := q.AsInt64(); !ok {
+		klog.V(4).InfoS("node resource's quantity is not int64 and will be rounded",
+			"node", node.Name, "resource", name, "original", *q, "rounded", q.Value())
+		q.Set(q.Value())
+	}
+	node.Status.Capacity[name] = *q
+	node.Status.Allocatable[name] = *q
+}
+
+// getPodMetricUsage gets pod usage from the PodMetricInfo
+func getPodMetricUsage(info *slov1alpha1.PodMetricInfo) corev1.ResourceList {
+	return getResourceListForCPUAndMemory(info.PodUsage.ResourceList)
+}
+
+// getPodNUMARequestAndUsage returns the pod request and usage on each NUMA nodes.
+// It averages the metrics over all sharepools when the pod does not allocate any sharepool or use all sharepools.
+func getPodNUMARequestAndUsage(pod *corev1.Pod, podRequest, podUsage corev1.ResourceList, numaNum int) ([]corev1.ResourceList, []corev1.ResourceList) {
```

Comment:
The code in this file has a lot of calculation logic, but it lacks unit-tests.

Reply:
UT added